converting into templates

ganesshkumar opened the pull request:
**Changes**
* Parsing the http headers manually. Removed the http-parser library as it depends on build-tools while installing in Dockerfile causing increased size of image. Also, the development of the [library](https://github.com/benoitc/http-parser) is very slow with unanswered issues over a long period of time.
* Converting to separate templates for python and python3

**Issues**
* I am running on `docker swarm` and during intallation I did `git checkout 0.6.5` before running `./deploy_stack.sh` .I tested nodejs-afterburn template and the templates in this PR. Everything is working file. However, if I do `git checkout 0.6.15` (the latest release) of openfaas/faas, both nodejs-afterburn and these python templates fail with `Unexpected EOF`. Has something changed significantly between these two tags that requires changes in all afterburn templates?

**Data**
- [Performance tests with fwatchdog](https://github.com/ganesshkumar/python-afterburn/wiki/Performance-Tests)
- [Usage and performance tests with openfaas 0.6.5](https://github.com/ganesshkumar/python-afterburn/wiki/Usage)

ganesshkumar committed:
converting into templates

viveksyngh commented:
@ganesshkumar I tried to test this code by creating a sample Fibonacci function. It works fine for the straight case but when function raises an exception it goes in some loop and it keeps printing below message in the container log. 

```
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
fib_afterburn.1.ooziv5bekco0@ip-172-31-15-222    | 2017/12/27 16:06:01 Error reading stderr: EOF
```

and it returns ```write |1: broken pipe``` this as the response to subsequent requests.

Please refer to attached screenshot.

<img width="1408" alt="screen shot 2017-12-27 at 9 44 56 pm" src="https://user-images.githubusercontent.com/3869193/34386574-5c25c440-eb4f-11e7-973e-03ff45ea481c.png">

ganesshkumar commented:
Oh yeah, I am not catching any exception. I will add exception handling and update the PR.

ganesshkumar committed:
returning 500 internal error when exception is thrown from the funciton

ganesshkumar commented:
I have updated the PR.

I test with this simple function. If you pass anything other than the string "pass", it will throw exception
```
def handle(st):
    if st == "pass":
        return "You said: " + st + ". That was so fast!"
    else:
        raise Exception("You must say pass")
```

**Testing via watchdog** 
Success
```
$ curl http://localhost:8081/ --data pass -i
HTTP/1.1 200 OK
Content-Length: 33
Content-Type: text/plain
Date: Thu, 28 Dec 2017 12:24:08 GMT

You said: pass. That was so fast!%
```
Exception
```
$ curl http://localhost:8081/ --data pas -i
HTTP/1.1 500 Internal Server Error
Content-Length: 33
Content-Type: text/plain
Date: Thu, 28 Dec 2017 12:09:44 GMT

Internal Error: You must say pass%
```

viveksyngh commented:
I have also tested. It is working fine now.

alexellis commented:
Hi guys - the error you are seeing "Error reading stderr: EOF" is from where the container process exits while in afterburn mode. We need to make sure the `HEALTHCHECK` (Swarm) / readiness probe (K8s) gets invalidated so the process is restarted.

Code is here: https://github.com/openfaas-incubator/of-watchdog

alexellis commented on template/python-afterburn/index.py:
```diff
@@ -0,0 +1,106 @@
+        raise KeyError('Content-Length must be present in the request')
+
+    content_length = int(headers['Content-Length'])
+    body = sys.stdin.read(content_length)
```
What should we do if the HTTP request is deliberately malformed? We could do this for instance with `netcat` - i.e. the body is empty or too short. What will happen?

ganesshkumar replied:
@alexellis @viveksyngh 

We can add a custom RequestMalformedError and use it to throw any error related to request. 

Upon catching it we can return 500 with body like `"Malformed Request: {}".format(' - '.join(error.args))`

alexellis commented:
@viveksyngh @ganesshkumar thank you guys these changes are starting to look really good 👍 🥇

ganesshkumar committed:
adding exception to http parser

ganesshkumar commented:
Hi guys, Happy New year :) 

I have adding few changes to this PR to handle the malformed requests.

1. I have added allowed http methods with which a http request should start with. 
```
HTTP_METHODS = ['POST']
```
As of now I have added only 'POST' as I notice only POST requests are received in afterburn(we can add other methods if needed). The method of the incoming request is checked against this array, to ensure it is valid request. So, if a request specifies Content-Length as 8 and sends only 4 characters in the body, the parser will take the first 4 characters from the next request and completes the first request. The second request will start failing as it will not start with the allowed http method.

2. I have added a Custom exception `MalformedRequestError` and receiving this exception, I am throwing '400 Bad Request'.

3. I am throwing any `MalformedRequestError` only after reading till the body(length mentioned as per the Content-Length of this request), to allow the next request start from the http method. 

4. I am ignoring any malformed http headers lines as of now, as long as Content-Length is present in the request. We may discuss on this as of what the right approach should be.

Note: Yet to check where to invalidate HEALTHCHECK (Swarm) / readiness probe.

alexellis commented:
@ganesshkumar @viveksyngh would be good to summarise on this and get it merged in 👍

alexellis commented on template/python3-afterburn/function/handler.py:
```diff
@@ -0,0 +1,2 @@
+def handle(st):
+    return "You said: " + st + ". That was so fast!"
```
> That was so fast!

Yes 👍 can we just have the regular message though for both?

alexellis commented on template/python3-afterburn/index.py:
```diff
@@ -0,0 +1,128 @@
+from function import handler
+
+SEPARATOR = "\r\n"
+HTTP_METHODS = ['POST']
```
We should support GET also

alexellis commented:
Please remind me why we dropped the parser library?
